### PR TITLE
Builtin/profiler: Replace builtin profiler

### PR DESCRIPTION
**Pull request by t4im**
Mod Profiler rewrite/merge of [profiler](https://github.com/t4im/profiler/) to replace the current minetest profiler.
Tracking issue #3849 

#### Changes
* Correct calculation of average (mean) values. (Also made undistorted by idleness.)
* Reduce instrumentation overhead.
* Updated and extended instrumentation
* Fix crashes related to missing parameters for the future. *(e.g. #4220)*
* Fix occasional DIV/0's.
* Prevent issues caused by timetravel (overflow, timejump, NTP corrections)
* Prevent modname clashes with internal names.
* Add `/profiler` command for profile-data management.
* Measure each instrumentation individually and label based on registration order.
* Allow labeling of ABM's and LBM's for easier classification.  
  Giving several ABM's or LBM's the same label will treat them as one.  
  Missing labels will be generated based on name or registration order.  
  *I've already added some labels to homedecor, pipeworks and technic last year; so you can already see that in action.*
* Improved table formatting.
* Measure profiler overhead
* Very configurable with settings menu integration.  
  (E.g. Skip instrumenting builtin)

#### Usage
Use the setting "profiler.load" to load the profiler at startup.

* `/profiler dump [filter]` - report statistics to STDOUT and debug.txt
* `/profiler print [filter]` - report statistics to in-game console
* `/profiler save [format [filter]]` - saves the runtime statistics to a file in your worldpath (`profile-<time>[-<filter>].<format>`)
  * `txt` (default) - same treetable format as used by the dump and print commands
  * `csv` - ready for spreadsheet import
  * `json` - useful for adhoc D3 visualizations
  * `json_pretty` - line wrapped and intended json for humans
  * `lua` - serialized lua table of the profile-data, for adhoc scripts  
* `/profiler reset` - reset all gathered profile data.
  This can be helpful to discard of any startup measurements that often spike during loading or to get more useful min-values.

 `[filter]` allows limiting the output of the data via substring/pattern matching against the modname.
Note: Serialized data structures might be subject to change with changed or added measurements.  
	csv might be the most stable, due to flat structure.

I've got some incomplete changes in stash, that i kept back to not mess with the stability of this version here. But if people are interested, i might add these at a later date (things like node instrumentation, call frequency, problem markers).

### CODE REVIEW

**Comment by Fixer-007** [comment]
This is very needed for debug

File: doc/lua_api.txt [hunk 1]
```diff
@@ -3402,6 +3405,9 @@ Definition tables
 ### LBM (LoadingBlockModifier) definition (`register_lbm`)
 
     {
+		label = "Upgrade legacy doors",
```
Comment (est31) [inline doc/lua_api.txt]: Don't use tabs in lua_api.txt.

File: builtin/profiler/init.lua [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+			return false, string.format(
+				"Usage: %s\n" ..
+				"Format can be one of txt, csv, lua, json (structures may be subject to change).",
+				param_usage
+			)
+		end
+	})
+end
+
+core.log("action", "Initialize mod profiler")
```
Comment (est31) [inline builtin/profiler/init.lua]: Is this needed?

File: builtin/profiler/sampling.lua [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+		}
+	}
+	stats_total = profile.stats_total
+
+	-- Provide access to the most recent profile.
+	sampler.profile = profile
+end
+
+---
+-- Log a measurment for the sampler to pick up later.
```
Comment (est31) [inline builtin/profiler/sampling.lua]: Is it really spelled "measurment"? I don't think it is

File: builtin/profiler/reporter.lua [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+		format
+	)
+
+	local output, err = io.open(path, "w")
+	if not output then
+		return false, "Saving of profile failed with: " .. err
+	end
+	local content, err = serialize_profile(profile, format, filter)
+	if not content then
+		return false, "Saving of profile failed with: " .. err
```
Comment (Zeno-) [inline builtin/profiler/reporter.lua]: Shouldn't `output` be closed before returning?

Comment (t4im) [inline builtin/profiler/reporter.lua]: Ah, yes, thanks. :-)

File: builtin/profiler/sampling.lua [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+end
+
+---
+-- Sample all logged measurements each server step.
+-- Like any globalstep function, this should not be too heavy,
+-- but does not add to the instrumentation overhead.
+--
+local function sample(dtime)
+	-- Rare, but happens and is currently of no informational value.
+	if logged_time == 0 then return end
```
Comment (est31) [inline builtin/profiler/sampling.lua]: @kwolekr doesn't like having the then part of an if clause on the same line.

**Comment by est31** [comment]
Other than that style issue (which appears to be existing in other parts as well), I'd give it a :+1:

**Comment by t4im** [comment]
Style issue is addressed. I'll pay attention to that next time.

**Comment by t4im** [comment]
When people copy profiler tables into github, it shouldn't mess up things so much anymore:

 instrumentation                                         |    min µs |    max µs |    avg µs | min % | max % | avg %
-------------------------------------------------------- | --------- | --------- | --------- | ----- | ----- | ------
 unified_inventory:                                      |         1 |     10631 |        41 |   0.3 |  75.8 |  24.2
  - globalstep[1] .....................................  |         1 |       191 |        15 |   0.0 |  75.8 |  24.0


Not very pretty, but still better than the alternative non-monospace amagedon.

**Comment by Zeno-** [comment]
I like it!

**Comment by est31** [comment]
me too!

**Comment by Zeno-** [comment]
+1

**Comment by t4im** [comment]
Rebased against master, squashed a few commits, edited a few messages.
No real changes since the last approval, [but a tiny fix found during testing](https://github.com/minetest/minetest/pull/4245/commits/7db4cc4bc9b8dad700ee4fc621fa32ecd31db75f#diff-c352a1dd10c1b6133672bd3dd435d984R109) (extra 'q' to not lose header entry)

File: builtin/settingtypes.txt [hunk 6]
```diff
@@ -1163,6 +1154,51 @@ secure.trusted_mods (Trusted mods) string
...
+
+[**Profiling]
+#    Load the game profiler to collect game profiling data.
+#    Provides a /profiler command to access the compiled profile.
+#    Useful for mod developers and server operators.
+profiler.load (Load the game profiler) bool false
+
+#    The default format in which profiles are being saved,
+#    when calling `/profiler save [format]` without format.
+profiler.default_report_format (Default report format) enum txt,csv,lua,json,json_pretty
```
Comment (est31) [inline builtin/settingtypes.txt]: Maybe add a default for this enum?

**Comment by est31** [comment]
+1

**Pull request merged**